### ISSUE TITLE
NodeUnpublish is called *after* a pod is deleted from API server

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#101281]
#### What happened:
Issue #101279 revealed interesting behavior of kubelet - NodeUnpublish is called few hundred milliseconds **after** a pod is deleted from the API server.

The test:

1. It [writes some data to a PV/PVC](https://github.com/kubernetes/kubernetes/blob/c555b23f1c84bcdd1d87a7ae831675281f5c0c94/test/e2e/storage/testsuites/snapshottable.go#L146-L148) and inside `RunInPodWithVolume` it [waits for the writer pod to disappear](https://github.com/kubernetes/kubernetes/blob/c555b23f1c84bcdd1d87a7ae831675281f5c0c94/test/e2e/storage/testsuites/provisioning.go#L751).
2. It takes snapshot of the PV/PVC.

On the CSI driver side I can see that CreateSnapshot is called *before* NodeUnpublish. How is it possible? I checked, the writer pod is really deleted at the end of step 1.

I was under impression that NodeUnpublish must always succeed before a pod can be deleted.

See this flake in GCP PD CSI driver: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/sigs.k8s.io_gcp-compute-persistent-disk-csi-driver/733/pull-gcp-compute-persistent-disk-csi-driver-kubernetes-integration/1375194813174910976

On the node:
```
21:47:09.627074       1 utils.go:60] /csi.v1.Node/NodeStageVolume called
21:47:10.423483       1 node.go:326] NodeStageVolume succeeded
21:47:10.436414       1 utils.go:60] /csi.v1.Node/NodePublishVolume called
21:47:10.470472       1 node.go:208] NodePublishVolume succeeded

21:47:13.978021       1 utils.go:60] /csi.v1.Node/NodeUnpublishVolume called
21:47:13.981197       1 node.go:243] NodeUnpublishVolume succeeded
21:47:57.506513       1 utils.go:60] /csi.v1.Node/NodeUnstageVolume called
...
```

In the controller pod:
```
21:47:08.918265       1 controller.go:377] ControllerPublishVolume succeeded
21:47:12.715796       1 utils.go:60] /csi.v1.Controller/CreateSnapshot called
```

-> `CreateSnapshot` was called 1 second before `NodeUnpublishVolume`. I saw the same behavior with AWS EBS driver and with both controller + node pods on the same node (so there can't be any clock skew).

Since https://github.com/kubernetes/kubernetes/pull/101280 waits for `NodeUnstage`, it will fix also this issue, maybe we don't need to worry about it at all.

#### Environment:
- Kubernetes version (use `kubectl version`): v1.21

cc @msau42 @jingxu97 @gnufied

**Comment by msau42** [comment]
That's very surprising. We check mountedvolumes from ASW first, it should have found it there: https://github.com/kubernetes/kubernetes/blob/6ede5ca95f78478fa627ecfea8136e0dff34436b/pkg/kubelet/kubelet_volumes.go#L55

But even assuming that, once it's unpublished, the pod can be deleted. We don't wait for unstage which is where the flush is done.

**Comment by gnufied** [comment]
yeah this is very surprising indeed. Can we rule out minor timing issue in timestamps logged in those log files?

**Comment by jsafrane** [comment]
For AWS I ran everything on a single node, therefore `klog.Infof` should add correct timestamp to logs. They may be buffered and flushed later, but the timestamp on the log line should be accurate.

**Comment by jsafrane** [comment]
(Not to derail this conversation, I noticed for AWS EBS it's not enough to unmount / unstage, the volume must be detached to get everything really written to the backend. But it's very rare, I saw it only once in ~100 attempts).

**Comment by jsafrane** [comment]
I think it could be caused by `RestartPolicy: v1.RestartPolicyNever` and the write command to finish quickly - maybe a finished pod is deleted quicker than a pod that still runs.
